expected_codes_by_category = {'Population': {'slug': 'population', 'codes': {'O3_B1': {'list': ['O3', 'B1'], 'details': [{'O3': {'type': 'oncogenic', 'category': 'population', 'options': ['MO'], 'description': 'Absent or very rare in population database (e.g. gnomAD)', 'annotations': ['gnomad']}}, {'B1': {'type': 'benign', 'category': 'population', 'options': ['SU', 'MO', 'ST', 'VS', 'SA'], 'description': 'Variant is present at a high level in a general population database (e.g. gnomAD)', 'annotations': ['gnomad']}}], 'dropdown': [{'value': 'O3_PE|B1_PE', 'text': 'Pending'}, {'value': 'O3_NA|B1_NA', 'text': 'Not applied'}, {'value': 'O3_MO|B1_NA', 'text': 'O3 Moderate (+2)'}, {'value': 'O3_NA|B1_SU', 'text': 'B1 Supporting (-1)'}, {'value': 'O3_NA|B1_MO', 'text': 'B1 Moderate (-2)'}, {'value': 'O3_NA|B1_ST', 'text': 'B1 Strong (-4)'}, {'value': 'O3_NA|B1_SA', 'text': 'B1 Standalone (-100)'}, {'value': 'O3_NA|B1_VS', 'text': 'B1 Very Strong (-8)'}], 'value': 'O3_PE|B1_PE', 'annotations': ['gnomad'], 'all_checks': ['Pending']}}, 'applied_codes': [], 'complete': False}, 'Mode Of Action': {'slug': 'mode-of-action', 'codes': {'B2': {'list': ['B2'], 'details': [{'B2': {'type': 'benign', 'category': 'mode_of_action', 'options': ['SA'], 'description': 'Variant does not fit the mode of action of the gene for the cancer under investigation', 'annotations': [None]}}], 'dropdown': [{'value': 'B2_PE', 'text': 'Pending'}, {'value': 'B2_NA', 'text': 'Not applied'}, {'value': 'B2_SA', 'text': 'B2 Standalone (-100)'}], 'value': 'B2_PE', 'annotations': [None], 'all_checks': ['Pending']}, 'O2': {'list': ['O2'], 'details': [{'O2': {'type': 'oncogenic', 'category': 'mode_of_action', 'options': ['SU', 'MO', 'ST', 'VS', 'SA'], 'description': 'Null variant in a tumour surpressor gene', 'annotations': [None]}}], 'dropdown': [{'value': 'O2_PE', 'text': 'Pending'}, {'value': 'O2_NA', 'text': 'Not applied'}, {'value': 'O2_VS', 'text': 'O2 Very Strong (+8)'}, {'value': 'O2_ST', 'text': 'O2 Strong (+4)'}, {'value': 'O2_MO', 'text': 'O2 Moderate (+2)'}, {'value': 'O2_SU', 'text': 'O2 Supporting (+1)'}, {'value': 'O2_SA', 'text': 'O2 Standalone (+100)'}], 'value': 'O2_PE', 'annotations': [None], 'all_checks': ['Pending']}}, 'applied_codes': [], 'complete': False}, 'Cancer Databases': {'slug': 'cancer-databases', 'codes': {'O4': {'list': ['O4'], 'details': [{'O4': {'type': 'oncogenic', 'category': 'cancer_databases', 'options': ['SU', 'MO', 'ST'], 'description': 'Enriched in a somatic variant database (compared to prevalence in controls)', 'annotations': [None]}}], 'dropdown': [{'value': 'O4_PE', 'text': 'Pending'}, {'value': 'O4_NA', 'text': 'Not applied'}, {'value': 'O4_ST', 'text': 'O4 Strong (+4)'}, {'value': 'O4_MO', 'text': 'O4 Moderate (+2)'}, {'value': 'O4_SU', 'text': 'O4 Supporting (+1)'}], 'value': 'O4_PE', 'annotations': [None], 'all_checks': ['Pending']}}, 'applied_codes': [], 'complete': False}, 'Effect On Protein': {'slug': 'effect-on-protein', 'codes': {'B4': {'list': ['B4'], 'details': [{'B4': {'type': 'benign', 'category': 'effect_on_protein', 'options': ['SU'], 'description': 'Synonymous/ intronic variants (outside canonical splice sites)', 'annotations': [None]}}], 'dropdown': [{'value': 'B4_PE', 'text': 'Pending'}, {'value': 'B4_NA', 'text': 'Not applied'}, {'value': 'B4_SU', 'text': 'B4 Supporting (-1)'}], 'value': 'B4_PE', 'annotations': [None], 'all_checks': ['Pending']}, 'B5': {'list': ['B5'], 'details': [{'B5': {'type': 'benign', 'category': 'effect_on_protein', 'options': ['SU'], 'description': 'In-frame deletion in a repetitive region with unknown function', 'annotations': [None]}}], 'dropdown': [{'value': 'B5_PE', 'text': 'Pending'}, {'value': 'B5_NA', 'text': 'Not applied'}, {'value': 'B5_SU', 'text': 'B5 Supporting (-1)'}], 'value': 'B5_PE', 'annotations': [None], 'all_checks': ['Pending']}, 'O5': {'list': ['O5'], 'details': [{'O5': {'type': 'oncogenic', 'category': 'effect_on_protein', 'options': ['MO', 'SU'], 'description': 'Missense at an amino acid where a different missense variant has previously been determined to be oncogenic, where LOF is a know mechanism of oncogenicity', 'annotations': [None]}}], 'dropdown': [{'value': 'O5_PE', 'text': 'Pending'}, {'value': 'O5_NA', 'text': 'Not applied'}, {'value': 'O5_MO', 'text': 'O5 Moderate (+2)'}, {'value': 'O5_SU', 'text': 'O5 Supporting (+1)'}], 'value': 'O5_PE', 'annotations': [None], 'all_checks': ['Pending']}, 'O7': {'list': ['O7'], 'details': [{'O7': {'type': 'oncogenic', 'category': 'effect_on_protein', 'options': ['ST', 'MO', 'SU'], 'description': 'Located in a mutational hotspot and/or critical and well-established functional domain (e.g. active site of an enzyme), without benign variation', 'annotations': [None]}}], 'dropdown': [{'value': 'O7_PE', 'text': 'Pending'}, {'value': 'O7_NA', 'text': 'Not applied'}, {'value': 'O7_ST', 'text': 'O7 Strong (+4)'}, {'value': 'O7_MO', 'text': 'O7 Moderate (+2)'}, {'value': 'O7_SU', 'text': 'O7 Supporting (+1)'}], 'value': 'O7_PE', 'annotations': [None], 'all_checks': ['Pending']}, 'O8': {'list': ['O8'], 'details': [{'O8': {'type': 'oncogenic', 'category': 'effect_on_protein', 'options': ['MO', 'SU'], 'description': 'Protein length change due to in-frame deletions/insertions in a non-repeat region or stop-loss variants', 'annotations': [None]}}], 'dropdown': [{'value': 'O8_PE', 'text': 'Pending'}, {'value': 'O8_NA', 'text': 'Not applied'}, {'value': 'O8_MO', 'text': 'O8 Moderate (+2)'}, {'value': 'O8_SU', 'text': 'O8 Supporting (+1)'}], 'value': 'O8_PE', 'annotations': [None], 'all_checks': ['Pending']}}, 'applied_codes': [], 'complete': False}, 'Computational': {'slug': 'computational', 'codes': {'O6_B3': {'list': ['O6', 'B3'], 'details': [{'O6': {'type': 'oncogenic', 'category': 'computational', 'options': ['SU'], 'description': 'Computational evidence supports a deleterious effect on the gene or gene product (conservation, evolutionary, splicing impact etc)', 'annotations': [None]}}, {'B3': {'type': 'benign', 'category': 'computational', 'options': ['SU'], 'description': 'Computational evidence DOES NOT support a deleterious effect on the gene product (conservation, evolutionary, splicing impact etc)', 'annotations': [None]}}], 'dropdown': [{'value': 'O6_PE|B3_PE', 'text': 'Pending'}, {'value': 'O6_NA|B3_NA', 'text': 'Not applied'}, {'value': 'O6_SU|B3_NA', 'text': 'O6 Supporting (+1)'}, {'value': 'O6_NA|B3_SU', 'text': 'B3 Supporting (-1)'}], 'value': 'O6_PE|B3_PE', 'annotations': [None], 'all_checks': ['Pending']}}, 'applied_codes': [], 'complete': False}, 'Phenotype': {'slug': 'phenotype', 'codes': {'O10_B7': {'list': ['O10', 'B7'], 'details': [{'O10': {'type': 'oncogenic', 'category': 'phenotype', 'options': ['MO', 'SU'], 'description': 'Tumour phenotype', 'annotations': [None]}}, {'B7': {'type': 'benign', 'category': 'phenotype', 'options': ['MO', 'SU'], 'description': 'Tumour phenotype', 'annotations': [None]}}], 'dropdown': [{'value': 'O10_PE|B7_PE', 'text': 'Pending'}, {'value': 'O10_NA|B7_NA', 'text': 'Not applied'}, {'value': 'O10_MO|B7_NA', 'text': 'O10 Moderate (+2)'}, {'value': 'O10_SU|B7_NA', 'text': 'O10 Supporting (+1)'}, {'value': 'O10_NA|B7_SU', 'text': 'B7 Supporting (-1)'}, {'value': 'O10_NA|B7_MO', 'text': 'B7 Moderate (-2)'}], 'value': 'O10_PE|B7_PE', 'annotations': [None], 'all_checks': ['Pending']}}, 'applied_codes': [], 'complete': False}}

expected_codes_dict = {'O3': {'type': 'oncogenic', 'category': 'population', 'options': ['MO'], 'description': 'Absent or very rare in population database (e.g. gnomAD)', 'annotations': ['gnomad']}, 'B1': {'type': 'benign', 'category': 'population', 'options': ['SU', 'MO', 'ST', 'VS', 'SA'], 'description': 'Variant is present at a high level in a general population database (e.g. gnomAD)', 'annotations': ['gnomad']}, 'O2': {'type': 'oncogenic', 'category': 'mode_of_action', 'options': ['SU', 'MO', 'ST', 'VS', 'SA'], 'description': 'Null variant in a tumour surpressor gene', 'annotations': [None]}, 'B2': {'type': 'benign', 'category': 'mode_of_action', 'options': ['SA'], 'description': 'Variant does not fit the mode of action of the gene for the cancer under investigation', 'annotations': [None]}, 'O4': {'type': 'oncogenic', 'category': 'cancer_databases', 'options': ['SU', 'MO', 'ST'], 'description': 'Enriched in a somatic variant database (compared to prevalence in controls)', 'annotations': [None]}, 'O5': {'type': 'oncogenic', 'category': 'effect_on_protein', 'options': ['MO', 'SU'], 'description': 'Missense at an amino acid where a different missense variant has previously been determined to be oncogenic, where LOF is a know mechanism of oncogenicity', 'annotations': [None]}, 'O7': {'type': 'oncogenic', 'category': 'effect_on_protein', 'options': ['ST', 'MO', 'SU'], 'description': 'Located in a mutational hotspot and/or critical and well-established functional domain (e.g. active site of an enzyme), without benign variation', 'annotations': [None]}, 'O8': {'type': 'oncogenic', 'category': 'effect_on_protein', 'options': ['MO', 'SU'], 'description': 'Protein length change due to in-frame deletions/insertions in a non-repeat region or stop-loss variants', 'annotations': [None]}, 'B4': {'type': 'benign', 'category': 'effect_on_protein', 'options': ['SU'], 'description': 'Synonymous/ intronic variants (outside canonical splice sites)', 'annotations': [None]}, 'B5': {'type': 'benign', 'category': 'effect_on_protein', 'options': ['SU'], 'description': 'In-frame deletion in a repetitive region with unknown function', 'annotations': [None]}, 'O6': {'type': 'oncogenic', 'category': 'computational', 'options': ['SU'], 'description': 'Computational evidence supports a deleterious effect on the gene or gene product (conservation, evolutionary, splicing impact etc)', 'annotations': [None]}, 'B3': {'type': 'benign', 'category': 'computational', 'options': ['SU'], 'description': 'Computational evidence DOES NOT support a deleterious effect on the gene product (conservation, evolutionary, splicing impact etc)', 'annotations': [None]}, 'O9': {'type': 'oncogenic', 'category': 'functional_studies', 'options': ['VS', 'ST', 'MO', 'SU'], 'description': 'Well-established in vitro or in vivo functional studies demonstrating a functionally abnormal result consistent with the mechanism of disease', 'annotations': [None]}, 'B6': {'type': 'benign', 'category': 'functional_studies', 'options': ['ST', 'MO', 'SU'], 'description': 'Well-established in vitro or in vivo functional studies demonstrating no damaging effect on protein function or RNA splicing', 'annotations': [None]}, 'O10': {'type': 'oncogenic', 'category': 'phenotype', 'options': ['MO', 'SU'], 'description': 'Tumour phenotype', 'annotations': [None]}, 'B7': {'type': 'benign', 'category': 'phenotype', 'options': ['MO', 'SU'], 'description': 'Tumour phenotype', 'annotations': [None]}}
